### ISSUE TITLE
BLAS dot implementation

### ISSUE THREAD

**Issue by khaotik** [Theano/libgpuarray#292]
TL;DR There should be a way to call vector-vector dot.

Consider the following theano code:
```
T = theano.tensor
x = T.matrix()
y, _ = th.scan(fn=lambda x : T.dot(x,x), sequences=x)
fn = theano.function([x], y)
x_val = numpy.random.rand(1000, 10000).astype(numpy.float32)
timeit( fn(x_val) )
```

On my laptop, the code runs 3.4 sec with the new backend, and 0.85 sec on the old backend with a [small patch](https://github.com/Theano/Theano/pull/5257). Theano don't have shape information at compile time, so the above code was converted into `gemv`, which is much slower than a vector-vector dot on GPU. This is one of the major reasons for the [huge GPU scan overhead](https://github.com/Theano/Theano/issues/1168).

Was about to make a PR like the one for the old backend. I'm a bit unsure what's the best approach to address this problem with the new backend:

1. Modify the `gemv` code in `libgpuarray` to detect row-vector matrix. This is the easiest, however rather hackish/ugly/shortsighted solution. The code must test `alpha==1.0 && beta==0.0` at runtime. This has to be done for all CUDA/OCL/\<potential future api\> implementation.

2. Add `dot` function in `libgpuarray`'s BLAS API, then modify `c_code` in Theano to do the shape detection. Cleaner than the 1st approach. The problems is `libgpuarray` will have to use a whole array object to represent a single dot product result -> another overhead.

3. Make Theano do better shape inference or scan optimization. The best permanent solution, with the potential to bring many other optimizations possible. This might get complicated ...

**Comment by nouiz** [comment]
Thanks for investigating. What is the time in the old back-end without the
patch? Just to make sure it is about 3.4s as in the new back-end. Just to
make sure we don't have an extra overhead in the new back-end. I don't
expect one, but as you are working on this, double checking that would be
great.

3. won't always work. Take the example when the full Theano graph is just a
dot. Then we can't infer any shapes. If the user pass a row and a col for
both matrices, it would be great to still have the optimized version. So
this mean we need a quick check of shapes at run time.

Personally, I would go with 1. It is simple to use and do the work. In
fact, I find it sad that cublas don't do that optimization itself! I don't
find that solution hackish/... I find it as a simple solution that is
clean. It respect the interface and what the user wanted (compute a gemv).
It just does it in a faster way.

@abergeron, do you have objections to approch 1?

**Comment by abergeron** [comment]
I think approach 2 is better overall since we wouldn't have to reproduce the "fix" across all blas bindings which means less work.  We would have to add a binding for dot(), but I don't think that is a bad thing.

The overhead of having a full array structure for a single value will be there anyway so there is not much to lose there.

**Comment by nouiz** [comment]
ok

**Comment by khaotik** [comment]
@nouiz Unfortunately, there seems to be more overhead for the new backend :(

[Link](https://gist.github.com/khaotik/9c7c72e562a919e241c9a09522fc4df5) to my test script.

Unpatched:
```
khaotik@KKPC:~/WKSP_DL/dev/better_scan$ ./run_cuda.sh 
Mapped name None to device cuda: GeForce GT 650M
Using device cuda
scan with dot: 3.383239984512329
scan with sum: 0.4118797779083252
full unrolled: 0.0891871452331543
>>> 
khaotik@KKPC:~/WKSP_DL/dev/better_scan$ ./run_gpu.sh 
Using gpu device 0: GeForce GT 650M (CNMeM is enabled with initial size: 45.0% of memory, cuDNN 5105)
Using device gpu
scan with dot: 3.255945587158203
scan with sum: 0.32773518562316895
full unrolled: 0.07604336738586426
>>> 
khaotik@KKPC:~/WKSP_DL/dev/better_scan$ ./run_cpu.sh 
Using device cpu
scan with dot: 0.25354480743408203
scan with sum: 0.1973247528076172
full unrolled: 0.1682417392730713
```

Patched old backend:
```
khaotik@KKPC:~/WKSP_DL/dev/better_scan$ ./run_gpu.sh 
Using gpu device 0: GeForce GT 650M (CNMeM is enabled with initial size: 45.0% of memory, cuDNN 5105)
Using device gpu
scan with dot: 0.8559339046478271
scan with sum: 0.34679555892944336
full unrolled: 0.07688093185424805
```

Maybe test this on other GPUs/OpenCL devices?
